sumologic/resource_sumologic_hierarchy.go

```
@@ -0,0 +1,326 @@
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringLenBetween(1, maxHierarchyNameLen),
+			},
+
+			"filter": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: getFilterSchema(),
+				},
+			},
+
+			"level": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: getLevelSchema(1),
+				},
+			},
+		},
+	}
+}
+
+func getFilterSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"key": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringLenBetween(1, maxFilterKeyLen),
+		},
+		"value": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringLenBetween(1, maxFilterValueLen),
+		},
+	}
+}
+
+func getLevelSchema(currentDepth int) map[string]*schema.Schema {
+	levelSchema := map[string]*schema.Schema{
+		"entity_type": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringLenBetween(1, maxEntityTypeLen),
+		},
+	}
+
+	var nextLevelsWithConditionsSchema map[string]*schema.Schema
+	if currentDepth < maxDepth {
+		nextLevelsWithConditionsSchema = map[string]*schema.Schema{
+			"next_levels_with_conditions": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: maxWidth,
+				Elem: &schema.Resource{
+					Schema: getLevelWithConditionSchema(currentDepth + 1),
+				},
+			},
+		}
+	} else {
+		nextLevelsWithConditionsSchema = map[string]*schema.Schema{
+			"next_levels_with_conditions": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 0,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		}
+	}
+	for k, v := range nextLevelsWithConditionsSchema {
+		levelSchema[k] = v
+	}
+
+	if currentDepth < maxDepth {
```

What is the behavior when depth > maxDepth? Do we silently drop levels after max depth is reached?

---

If I understand correctly the provider will not accept such hierarchy because of unexpected field being present

---

I think what you said makes sense. Can you test this locally by setting `maxDepth` to a lower value, like 2?

---

```
- Unsupported block type: Blocks of type "next_levels_with_conditions" are not expected here. Did you mean to define argument "next_levels_with_conditions"? If so, use the equals sign to assign it a value.
- Unsupported block type: Blocks of type "next_level" are not expected here.
```